lib/Auth/Process/LogUser.php

```diff
@@ -0,0 +1,224 @@
+
+        $userAttributes = $this->getUserAttributes($state);
+
+        $logJson = '
+            {
+                "ID": "' . uniqid() . '",
+                "IDP": "' . $idp . '",
+                "SP": "' . $spEntityId . '",' .
+                $userAttributes .
+                '"Time": "' . $datetime . '"
+            }';
+
+        $item = $marshaler->marshalJson($logJson);
+
+        $params = [
+            'TableName' => $this->dynamoLogTable,
+            'Item' => $item,
+        ];
+
+        try {
+            $result = $dynamodb->putItem($params);
+        } catch (DynamoDbException $e) {
+            echo "Unable to add item:\n";
+            echo $e->getMessage() . "\n";
+        }
+    }
+
+    private function configsAreValid() {
+        $msg = ' config value not provided to LogUser.';
+        if (empty($this->dynamoEndpoint)) {
+            SimpleSAML\Logger::error(self::DYNAMO_ENDPOINT_KEY . $msg);
+            return false;
+        }
+
+        if (empty($this->dynamoRegion)) {
+            SimpleSAML\Logger::error(self::DYNAMO_REGION_KEY . $msg);
+            return false;
+        }
+
+        if (empty($this->dynamoLogTable)) {
+            SimpleSAML\Logger::error(self::DYNAMO_LOG_TABLE_KEY . $msg);
+            return false;
+        }
+
+        return true;
+    }
+
+    private function getIdp(&$state) {
+        if (empty($state[self::IDP_KEY])) {
+            return 'No IDP available';
+        }
+
+        $samlIDP = $state[self::IDP_KEY];
+
+        // Get the potential IDPs from idp remote metadata
+        $metadataPath = __DIR__ . '/../../../../../metadata';
+
+        // If a unit test sends a different metadataPath, use it
+        if (isset($state['metadataPath'])) {
+            $metadataPath = $state['metadataPath'];
+        }
+        $idpEntries = \Sil\SspUtils\Metadata::getIdpMetadataEntries($metadataPath);
+
+        // Get the IDPNamespace or else just use the IDP's entity ID
+        $idpEntry = $idpEntries[$samlIDP];
+
+        // If the IDPNamespace entry is close to being alphanumeric, use it
+        if (isset($idpEntry[self::IDP_CODE_KEY]) && is_string($idpEntry[self::IDP_CODE_KEY])) {
+            if ( preg_match("/^[A-Za-z0-9_-]+$/", $idpEntry[self::IDP_CODE_KEY])) {
+                return $idpEntry[self::IDP_CODE_KEY];
+            }
+        }
+        // Default, just use the idp's entity ID
+        return $samlIDP;
+    }
+
+    // Get the current user's common name attribute and/or eduPersonPrincipalName and/or employeeNumber
+    private function getUserAttributes($state) {
+        $attributes = $state['Attributes'];
+
+        $oidForCn = 'urn:oid:2.5.4.3';
+        $cnKey = 'cn';
+        
+        $oidForEduPersonPrincipalName = 'urn:oid:1.3.6.1.4.1.5923.1.1.1.6';
+        $eduPersonPrincipalNameKey = 'eduPersonPrincipalName';
+
+        $oidForEmployeeNumber = 'urn:oid:2.16.840.1.113730.3.1.3';
+        $employeeNumberKey = 'employeeNumber';
+        
+
+        $cn = '';
```

Comment:
Similar to the above situation, this...
```php
        $cn = '';
        if (!empty($attributes[$oidForCn])) {
            $cn =  $attributes[$oidForCn][0];
        } else if (!empty($attributes[$cnKey])) {
            $cn =  $attributes[$cnKey][0];
        }
```
... could be simplified to something like this:
```php
        $cn = $attributes[$oidForCn][0] ?? $attributes[$cnKey][0] ?? '';
```
That does result in a lot going on in that one line, though, so it's kind of a tradeoff. Feel free to go with whichever you like.

Reply:
Alternatively, you could extract this to another method... something along these lines:
```php
    $cn = $this->getAttributeFrom($attributes, $cnKey, $oidForCn);
    
    // ...
}

private function getAttributeFrom($attributes, $friendlyKey, $oidKey)
{
    $result = '';
    if (! empty($attributes[$oidKey])) {
        $result = $attributes[$oidKey][0];
    } else if (! empty($attributes[$friendlyKey])) {
        $result = $attributes[$friendlyKey][0];
    }
    return $result;
}
```